### ISSUE TITLE
Training with class weights drops the accuracy to 0

### ISSUE THREAD

**Issue by mitalbert** [TimoSaemann/ENet#22]
Hi. I'm trying to reproduce the results with class weights.

Would you please say have you trained the ENet with class weights successfully?

This happens every time(at different iterations) I try to train **the decoder** with class weights:

```
I0720 02:37:59.913910 13471 solver.cpp:228] Iteration 3140, loss = 0.604358
I0720 02:37:59.913946 13471 solver.cpp:244]     Train net output #0: accuracy = 0.490742
I0720 02:37:59.913954 13471 solver.cpp:244]     Train net output #1: loss = 0.604358 (* 1 = 0.604358 loss)
I0720 02:37:59.913959 13471 solver.cpp:244]     Train net output #2: per_class_accuracy = 0.582444
I0720 02:37:59.913961 13471 solver.cpp:244]     Train net output #3: per_class_accuracy = 0.479704
I0720 02:37:59.913964 13471 solver.cpp:244]     Train net output #4: per_class_accuracy = 0.330785
I0720 02:37:59.913969 13471 solver.cpp:244]     Train net output #5: per_class_accuracy = 0
I0720 02:37:59.913971 13471 solver.cpp:244]     Train net output #6: per_class_accuracy = 0.332091
I0720 02:37:59.913975 13471 solver.cpp:244]     Train net output #7: per_class_accuracy = 0.487523
I0720 02:37:59.913977 13471 solver.cpp:244]     Train net output #8: per_class_accuracy = 0.552928
I0720 02:37:59.913981 13471 solver.cpp:244]     Train net output #9: per_class_accuracy = 0.650392
I0720 02:37:59.913985 13471 solver.cpp:244]     Train net output #10: per_class_accuracy = 0.562122
I0720 02:37:59.913987 13471 solver.cpp:244]     Train net output #11: per_class_accuracy = 0
I0720 02:37:59.913991 13471 solver.cpp:244]     Train net output #12: per_class_accuracy = 0.976575
I0720 02:37:59.913995 13471 solver.cpp:244]     Train net output #13: per_class_accuracy = 0.738946
I0720 02:37:59.913997 13471 solver.cpp:244]     Train net output #14: per_class_accuracy = 0
I0720 02:37:59.914001 13471 solver.cpp:244]     Train net output #15: per_class_accuracy = 0.658808
I0720 02:37:59.914005 13471 solver.cpp:244]     Train net output #16: per_class_accuracy = 0
I0720 02:37:59.914011 13471 solver.cpp:244]     Train net output #17: per_class_accuracy = 0
I0720 02:37:59.914014 13471 solver.cpp:244]     Train net output #18: per_class_accuracy = 0
I0720 02:37:59.914034 13471 solver.cpp:244]     Train net output #19: per_class_accuracy = 0
I0720 02:37:59.914049 13471 solver.cpp:244]     Train net output #20: per_class_accuracy = 0
I0720 02:37:59.914054 13471 sgd_solver.cpp:106] Iteration 3140, lr = 0.005
I0720 02:38:23.496275 13471 solver.cpp:228] Iteration 3160, loss = 25.5314
I0720 02:38:23.496443 13471 solver.cpp:244]     Train net output #0: accuracy = 0.00153637
I0720 02:38:23.496454 13471 solver.cpp:244]     Train net output #1: loss = 25.5314 (* 1 = 25.5314 loss)
I0720 02:38:23.496459 13471 solver.cpp:244]     Train net output #2: per_class_accuracy = 0
I0720 02:38:23.496461 13471 solver.cpp:244]     Train net output #3: per_class_accuracy = 0
I0720 02:38:23.496464 13471 solver.cpp:244]     Train net output #4: per_class_accuracy = 0
I0720 02:38:23.496467 13471 solver.cpp:244]     Train net output #5: per_class_accuracy = 0
I0720 02:38:23.496471 13471 solver.cpp:244]     Train net output #6: per_class_accuracy = 0
I0720 02:38:23.496474 13471 solver.cpp:244]     Train net output #7: per_class_accuracy = 0
I0720 02:38:23.496477 13471 solver.cpp:244]     Train net output #8: per_class_accuracy = 0
I0720 02:38:23.496480 13471 solver.cpp:244]     Train net output #9: per_class_accuracy = 0
I0720 02:38:23.496484 13471 solver.cpp:244]     Train net output #10: per_class_accuracy = 0
I0720 02:38:23.496487 13471 solver.cpp:244]     Train net output #11: per_class_accuracy = 0
I0720 02:38:23.496490 13471 solver.cpp:244]     Train net output #12: per_class_accuracy = 0
I0720 02:38:23.496493 13471 solver.cpp:244]     Train net output #13: per_class_accuracy = 0
I0720 02:38:23.496496 13471 solver.cpp:244]     Train net output #14: per_class_accuracy = 0
I0720 02:38:23.496500 13471 solver.cpp:244]     Train net output #15: per_class_accuracy = 0
I0720 02:38:23.496505 13471 solver.cpp:244]     Train net output #16: per_class_accuracy = 0
I0720 02:38:23.496510 13471 solver.cpp:244]     Train net output #17: per_class_accuracy = 0
I0720 02:38:23.496512 13471 solver.cpp:244]     Train net output #18: per_class_accuracy = 0
I0720 02:38:23.496516 13471 solver.cpp:244]     Train net output #19: per_class_accuracy = 0
I0720 02:38:23.496520 13471 solver.cpp:244]     Train net output #20: per_class_accuracy = 1
I0720 02:38:23.496525 13471 sgd_solver.cpp:106] Iteration 3160, lr = 0.005
```
After removing the weights the training proceeded, however my aim is to train with class weights.

I tried to do everything as instructed. One thing I'm not sure about is the number of iterations **when training the encoder**. I did not see convergence after 45000 iterations(batch size 4). I also tried changing the learning rate until lr = 5e-08, however here is what I have at the "end" of encoder training:

```
lr = 5e-08
I0720 01:28:11.944630 12740 solver.cpp:228] Iteration 45700, loss = 0.264688
I0720 01:28:11.944663 12740 solver.cpp:244]     Train net output #0: accuracy = 0.755066
I0720 01:28:22.151648 12740 solver.cpp:228] Iteration 45720, loss = 0.20893
I0720 01:28:22.151680 12740 solver.cpp:244]     Train net output #0: accuracy = 0.769897
I0720 01:28:32.375144 12740 solver.cpp:228] Iteration 45740, loss = 0.574258
I0720 01:28:32.375293 12740 solver.cpp:244]     Train net output #0: accuracy = 0.628479
I0720 01:28:42.597137 12740 solver.cpp:228] Iteration 45760, loss = 0.559977
I0720 01:28:42.597182 12740 solver.cpp:244]     Train net output #0: accuracy = 0.703003
I0720 01:28:52.837726 12740 solver.cpp:228] Iteration 45780, loss = 0.27782
I0720 01:28:52.837759 12740 solver.cpp:244]     Train net output #0: accuracy = 0.550903
I0720 01:29:03.087321 12740 solver.cpp:228] Iteration 45800, loss = 0.331952
I0720 01:29:03.087358 12740 solver.cpp:244]     Train net output #0: accuracy = 0.74884

```
Would you please say, how do you determine the convergence when training the encoder(loss value and difference)? How many iterations did it take? What was the batch size? In case possible, would you please also provide the log file?

**Comment by mitalbert** [comment]
I just realized that calculating the class weights for the size 1024x2048 and then using these while resizing to 512x1024 can be the reason of this problems(especially low accuracy). I resized the images to 512x1024 and calculated the weights again:
```
1024x2048 vs 512x1024
0.0819	0.0713
0.4754	0.4287
0.1324	0.1195
1.5224	2.7831
1.5190	2.4387
2.4730	1.9900
8.1865	5.6761
5.2286	3.7375
0.1870	0.1704
1.4695	2.0416
0.6893	0.6807
1.9814	2.2165
7.8091	10.2256
0.4164	0.3892
1.3809	5.6482
1.1982	9.1413
0.6273	7.0231
5.3535	22.2056
4.0939	6.5786
```

As you can see the difference is in some cases too big. Can someone confirm this numbers?
I used xnview to resize the images and calculate_class_weighting.py (num_classes 19) to calculate.

**Comment by ASONG0506** [comment]
hi, could  you please tell me  how did you  solve the problem you mentioned above? I also encounter the same problem as yours when I am training the net on my own dataset, thank you very much!
![image](https://user-images.githubusercontent.com/26050577/53061558-f8aafd00-34f8-11e9-9061-be39824ff77b.png)
